### DISCUSSION TITLE
Delete-before-add semantics (AKA: conflicting effects)

### DISCUSSION THREAD

**Discussion by mikand** [aiplan4eu/unified-planning#191] [General]
This is to discuss the behavior of the UP when conflicting effects **in the same action** are declared (or appear in compilations - e.g. in grounding).

Generally, effects can be either conditional or non-conditional and can be assignments (e.g. "x gets the value 5") or increase/decrease (e.g. " decrease x by 5"). The current implementation of the UP implements the following behavior:

|  | Assign | Increase/Decrease | Conditional Assign | Conditional Increase/Decrease
|-|-|-|-|-
| **Assign** | Error in construction | Error in construction | Error in simulator if condition is true |  Error in simulator if condition is true
| **Increase/Decrease** | Error in construction | OK | Error in simulator if condition is true | OK
| **Conditional Assign** | Error in simulator if condition is true | Error in simulator if condition is true | Error in simulator if both conditions are true | Error in simulator if both conditions are true
| **Conditional Increase/Decrease** | Error in simulator if condition is true | OK | Error in simulator if both conditions are true | OK

Generally speaking, we semantically avoid more than two effects (at the same time in case of temporal planning) in the same action unless they are increase/decrease and we syntactically check that this is not written in the model only for the non-conditional case.

Now, in some planning semantics we have the "delete-before-add" semantics, where **for predicates only** it is allowed to have more than one assignment and this can emerge as a byproduct of some compilations and by grounding.

The key question here is to define (and document!) a semantics to be adopted by the UP (and exemplified by the simulator implementation) that:
- Makes sense :smile: 
- Allows the implementation of the classical compilations
- Is compatible with as many engines as possible

**Comment by mikand** [comment]
@roeger @hstairs Any idea on this topic? How do you like the table above?

**Comment by hstairs** [comment]
The table makes sense Andrea! 
Regarding the DBA, I would use that as it is what implemented in the majority of planners. But here @roeger  is more expert

**Comment by Framba-Luca** [comment]
| **Conflicting effects on the same Fluent** | **Boolean Assign** |                             **UserType Assign**                            |                                       **Conditional UserType Assign**                                       |                              **Numeric Assign**                             |                                       **Conditional Numeric Assign**                                       |          **Numeric Increase/Decrease**         |       **Conditional Numeric Increase/Decrease**      |
|:------------------------------------------:|:------------------:|:--------------------------------------------------------------------------:|:-----------------------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------:|:----------------------------------------------------------------------------------------------------------:|:----------------------------------------------:|:----------------------------------------------------:|
|             **Boolean Assign**             |  Add-After-Delete  |                                 Impossible                                 |                                                  Impossible                                                 |                                  Impossible                                 |                                                 Impossible                                                 |                   Impossible                   |                      Impossible                      |
|             **UserType Assign**            |                    | 1)Error at creation time? 2)Error at simulation  time if different values? |      If condition is True: 1)Error at simulation time? 2) Error at simulation time if different values?     |                                  Impossible                                 |                                                 Impossible                                                 |                   Impossible                   |                      Impossible                      |
|       **Conditional UserType Assign**      |                    |                                                                            | If both conditions are True: 1) Error at simulation time? 2) Error at simulation  time if different values? |                                  Impossible                                 |                                                 Impossible                                                 |                   Impossible                   |                      Impossible                      |
|             **Numeric Assign**             |                    |                                                                            |                                                                                                             | 1) Error at creation time? 2)Error at simulation  time if different values? |      If condition is True: 1)Error at simulation time 2) Error at simulation time if different values?     |             Error at creation time             |    If condition is True: Error at simulation time    |
|       **Conditional Numeric Assign**       |                    |                                                                            |                                                                                                             |                                                                             | If both conditions are True: 1) Error at simulation time? 2) Error at simulation  time if different value? | If condition is True: Error at simulation time | If both condition are True: Error at simulation time |
|        **Numeric Increase/Decrease**       |                    |                                                                            |                                                                                                             |                                                                             |                                                                                                            |                     Correct                    |                        Correct                       |
|  **Conditional Numeric Increase/Decrease** |                    |                                                                            |                                                                                                             |                                                                             |                                                                                                            |                                                |                        Correct                       |

This is the new table including the add-after-delete semantic. 
It's still to decide if we want to take route 1 or 2, understanding what happens when 2 numeric assign (for example) assign the same value.

The 1 is currently implemented.
